WearSpeakerSample/wear/src/main/java/com/example/android/wearable/speaker/MainStateHolder.kt

```diff
@@ -16,161 +16,215 @@
-                when (oldState) {
-                    is AppState.Ready,
-                    AppState.PlayingMusic,
-                    AppState.Recording -> {
-                        if (speakerIsSupported()) {
-                            AppState.PlayingVoice
-                        } else {
-                            showSpeakerNotSupported()
-                            AppState.Ready(transitionInstantly = false)
+                        // If we were already playing, transition back to ready
+                        AppState.PlayingVoice -> {
+                            appState = AppState.Ready(transitionInstantly = false)
                         }
                     }
-                    // If we were already playing, transition back to ready
-                    AppState.PlayingVoice -> AppState.Ready(transitionInstantly = false)
-                }
-            // We finished doing something, transition back to ready
-            AppAction.PlayMusicFinished,
-            AppAction.PlayRecordingFinished,
-            AppAction.RecordingFinished -> AppState.Ready(transitionInstantly = false)
-            AppAction.PermissionResultReturned ->
-                // Check if the user granted the permission
-                if (
-                    ContextCompat.checkSelfPermission(application, Manifest.permission.RECORD_AUDIO) ==
-                    PackageManager.PERMISSION_GRANTED
-                ) {
-                    // The user granted the permission, continue on to start recording
-                    AppState.Recording
-                } else {
-                    // We have confirmation now that the user denied the permission
-                    isPermissionDeniedMutableStateFlow.value = true
-                    AppState.Ready(transitionInstantly = false)
+                AppAction.PermissionResultReturned ->
+                    // Check if the user granted the permission
+                    if (
+                        ContextCompat.checkSelfPermission(activity, Manifest.permission.RECORD_AUDIO) ==
+                        PackageManager.PERMISSION_GRANTED
+                    ) {
+                        // The user granted the permission, continue on to start recording
+                        recordUpdatingState()
+                    } else {
+                        // We have confirmation now that the user denied the permission
+                        isPermissionDenied = true
+                        appState = AppState.Ready(transitionInstantly = false)
+                    }
+            }
+        }
+    }
+
+    /**
+     * A helper function to record, updating the progress state while recording.
+     *
+     * This requires the [Manifest.permission.RECORD_AUDIO] permission to run.
+     */
+    @RequiresPermission(Manifest.permission.RECORD_AUDIO)
+    private suspend fun recordUpdatingState() {
+        // We have the permission, we can start recording now
+        appState = AppState.Recording
```

Comment:
In these two last functions, the mutex isn't used. Use it or reconsider using `StateFlow`